src/storage/mod.rs

```diff
@@ -1595,83 +1641,100 @@ impl<E: Engine> Storage<E> {
+                        } else {
+                            Some(Key::from_encoded(end_key))
+                        };
+                        let pairs = if reverse {
+                            Self::reverse_raw_scan(
+                                &snapshot,
+                                &cf,
+                                &start_key,
+                                end_key,
+                                each_limit,
+                                &mut statistics,
+                                key_only,
+                            )?
+                        } else {
+                            Self::raw_scan(
+                                &snapshot,
+                                &cf,
+                                &start_key,
+                                end_key,
+                                each_limit,
+                                &mut statistics,
+                                key_only,
+                            )?
+                        };
+                        result.extend(pairs.into_iter());
+                    }
 
-                        thread_ctx.collect_read_flow(region_id, &statistics);
-                        thread_ctx.collect_key_reads(CMD, statistics.write.flow_stats.read_keys as u64);
-                        thread_ctx.collect_scan_count(CMD, &statistics);
+                    thread_ctx.collect_read_flow(region_id, &statistics);
+                    thread_ctx.collect_key_reads(CMD, statistics.write.flow_stats.read_keys as u64);
+                    thread_ctx.collect_scan_count(CMD, &statistics);
 
-                        Ok(result)
-                    })
-                    .and_then(move |res| {
-                        timer.observe_duration();
-                        let r = tx.send(res);
-                        if r.is_err() {
-                            warn!("future_callback: Failed to send result to the future rx, discarded.");
-                        }
-                        Ok(())
-                    })
-            }) {
-                fut.forget();
+                    Ok(result)
+                })
+                .and_then(move |res| {
+                    timer.observe_duration();
+                    let r = tx.send(res);
+                    if r.is_err() {
+                        warn!(
+                            "future_callback: Failed to send result to the future rx, discarded."
+                        );
+                    }
+                    Ok(())
+                })
+        }) {
+            Ok(res) => res.forget(),
+            Err(_e) => {
+                let _ = tx2.send(Error::SchedTooBusy);
             }
         };
 
-        let res = engine.async_snapshot(&ctx, cb);
-        future::result(res)
-            .map(|_| future.map_err(|e| EngineError::Other(box_err!(e))))
-            .flatten()
-            .map_err(|_| Error::SchedTooBusy)
+        let _res = engine.async_snapshot(&ctx, cb);
+
+        future
```

Comment:
Impressive! Could you also write a failpoint test (or normal tests are also fine!) to verify that this really works (I mean, the error is really returned when future pool is full)?